### PR TITLE
[fsdp] feat: integrate PrefixGrouper for GRPO training acceleration

### PR DESCRIPTION
**Pull request by kevssim**
### What does this PR do?

Integrate [PrefixGrouper](https://github.com/johncaged/PrefixGrouper) into verl's FSDP worker to accelerate GRPO training by reducing redundant prefix computations.

In GRPO training, each prompt is copied `G` times (rollout.n), leading to redundant self-attention computation on shared prefixes. PrefixGrouper decomposes this into **prefix self-attention + suffix concat-attention**, significantly reducing computation and memory usage.

**Key changes:**
- Add `use_prefix_grouper` config option in `ActorConfig`
- Implement PG forward path in `DataParallelPPOActor._forward_micro_batch`
- Add utility functions in `verl/trainer/ppo/prefix_grouper_utils.py`
- Add example scripts and documentation in `examples/prefix_grouper_examples/`

### Test

**Benchmark Results** (Qwen3-4B, 4×H800, `rollout.n=4`):

| Context Length | Metric | PG | No PG | Speedup |
|----------------|--------|-----|-------|---------|
| **4K** | `old_log_prob` | 1.31s | 1.70s | **1.30x** |
| | `update_actor` | 4.80s | 6.07s | **1.26x** |
| | `step` | 17.08s | 19.40s | **1.14x** |
| **8K** | `old_log_prob` | 1.69s | 2.63s | **1.56x** |
| | `update_actor` | 5.98s | 10.18s | **1.70x** |
| | `step` | 19.48s | 24.71s | **1.27x** |

timing_comparison_combined

As context length increases, the speedup becomes more pronounced.

### API and Usage Example

```python
# Enable PrefixGrouper in training config
actor_rollout_ref.actor.use_prefix_grouper=True
trainer.balance_batch=False  # Required: PG is incompatible with balance_batch
actor_rollout_ref.model.use_remove_padding=False  # Required: PG is incompatible with remove_padding
```

```bash
# Run example script
bash examples/prefix_grouper_examples/run_qwen3_pg.sh
```

### Design & Code Changes

**High-level Design:**

PrefixGrouper optimizes GRPO training by avoiding redundant computation on shared prefixes. When `rollout.n > 1`, multiple responses share the same prompt, but standard attention computes the prefix `n` times. PrefixGrouper decomposes this into:
1. **Prefix self-attention**: Compute once per unique prompt
2. **Suffix concat-attention**: Each response attends to the shared prefix output

### Design & Code Changes

**High-level Design:**

PrefixGrouper optimizes GRPO training by avoiding redundant computation on shared prefixes. When `rollout.n > 1`, multiple responses share the same prompt, but standard attention computes the prefix `n` times. PrefixGrouper decomposes this into:
1. **Prefix self-attention**: Compute once per unique prompt
2. **Suffix concat-attention**: Each response attends to the shared prefix output

**Code Changes:**

| File | Change |
|------|--------|
| `verl/workers/config/actor.py` | Add `use_prefix_grouper: bool = False` config option |
| `verl/trainer/config/actor/actor.yaml` | Add `use_prefix_grouper: false` default config |
| `verl/workers/actor/dp_actor.py` | (1) Add `self.use_prefix_grouper` and `self.use_dynamic_bsz` attributes in `__init__`; (2) Add PG forward path in `_forward_micro_batch` with lazy import and incompatibility checks; (3) Select extra keys (`prompts`, `response_mask`, `uid`) for PG in `compute_log_prob`; (4) Select extra keys (`prompts`, `uid`) for PG in `update_policy` |
| `verl/trainer/ppo/prefix_grouper_utils.py` | New file with: `build_position_ids_for_prefix_grouper()` for position encoding, `build_pg_from_micro_batch()` to construct PrefixGrouper from micro batch, `pg_forward()` to execute PG-optimized forward pass |
| `verl/workers/fsdp_workers.py` | Sync `use_prefix_grouper` config from actor to ref policy in `init_model` to ensure both use the same forward path |
| `verl/trainer/ppo/ray_trainer.py` | Add `ValueError` check for `use_prefix_grouper + balance_batch` incompatibility at initialization |
| `examples/prefix_grouper_examples/` | New directory with: `README.md` documentation, `run_qwen3_prefix_grouper.sh` example script, `qwen3/modeling_qwen3.py` modified model supporting PrefixGrouper |

### Limitations

- **FSDP worker only**: Megatron worker is not supported yet
- **Incompatible configurations:**
  - `use_dynamic_bsz=True`
  - `use_remove_padding=True` (Flash Attention V2 variable length)
  - `use_fused_kernels=True`
  - `use_ulysses_sp=True` (Ulysses sequence parallelism)
- **Model modification required**: The model must accept `prefix_grouper` argument in its `forward` method

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs). (Added [examples/prefix_grouper_examples/README.md](cci:7://file:///d:/workspace/verl-tpx/examples/prefix_grouper_examples/README.md:0:0-0:0))
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: PrefixGrouper requires modified model files and specific hardware setup, tested manually with benchmark results above.
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1).

### CODE REVIEW

File: verl/workers/actor/dp_actor.py [hunk 1]
```diff
@@ -99,6 +103,78 @@ def _forward_micro_batch(
             entropy: # (bs, response_len)
             log_probs: # (bs, response_len)
         """
+        # PrefixGrouper path for shared-prefix optimization
+        if self.use_prefix_grouper:
```
Comment (wuxibin89) [inline verl/workers/actor/dp_actor.py:107]: Better move #L106-177 to prefix_grouper_utils.py as a separate function and keep dp_actor.py concise.
```python
if self.use_prefix_grouper:
    from verl.trainer.ppo.prefix_grouper_utils import forward_micro_batch
    return forward_micro_batch(micro_batch, ...)
```

File: verl/trainer/ppo/ray_trainer.py [hunk 2]
```diff
@@ -343,6 +343,15 @@ def __init__(
         if self.config.algorithm.use_kl_in_reward:
             self.kl_ctrl_in_reward = core_algos.get_kl_controller(self.config.algorithm.kl_ctrl)
 
+        # PrefixGrouper requires data to be sorted by uid, which is incompatible with balance_batch
+        if self.config.actor_rollout_ref.actor.get("use_prefix_grouper", False) and self.config.trainer.get(
```
Comment (wuxibin89) [inline verl/trainer/ppo/ray_trainer.py]: Sequence balance is important to balance num of total tokens in micro batches across all dp groups, I'm a bit concerned that this incompatibility will diminish the benefits of prefix_grouper.

File: examples/prefix_grouper/qwen3/modeling_qwen3.py [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+
+    attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query.dtype)
+    attn_weights = nn.functional.dropout(attn_weights, p=dropout, training=module.training)
+    attn_output = torch.matmul(attn_weights, value_states)
+    attn_output = attn_output.transpose(1, 2).contiguous()
+
+    return attn_output, attn_weights
+
+
+class Qwen3Attention(nn.Module):
```
Comment (wuxibin89) [inline examples/prefix_grouper/qwen3/modeling_qwen3.py]: Is there's a uniform way to monkey patch transformer for all models? It's too verbose for user to patch each model independently.

For example, we have a uniform monkey patch for deepspeed ulyssess:
https://github.com/volcengine/verl/blob/main/verl/models/transformers/monkey_patch.py

File: verl/trainer/ppo/prefix_grouper_utils.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        prefix_grouper,
+        concat_input_ids,
+        attention_mask,
+        position_ids,
+        responses,
+        response_mask,
+    )
+
+
+def pg_forward(
```
Comment (wuxibin89) [inline verl/trainer/ppo/prefix_grouper_utils.py:103]: Does prefix_grouper forward compatible with ulyssess sequence parallel?

Comment (kevssim) [reply verl/trainer/ppo/prefix_grouper_utils.py:103]: Currently, prefix_grouper is not compatible with Ulysses sequence parallel.

PrefixGrouper Relies on computing the full prefix once and reusing it for multiple suffixes. When the sequence is split across devices, we can't efficiently share the prefix computation across grouped samples.

The choice between them depends on your workload:
- Ulysses SP: Better for very long sequences where sequence length is the bottleneck
- PrefixGrouper: Better when you have larger n (e.g., 8+) and high prefix/suffix ratio (long prompts, short responses)

File: verl/trainer/ppo/ray_trainer.py (hunk 2 above)
Comment (kevssim) [reply verl/trainer/ppo/ray_trainer.py]: Now we address the incompatibility issue by implementing group level sequence balancing.

**Comment by wuxibin89** [comment]
Please format code: https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting

**Comment by kevssim** [comment]
> Please format code: https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting

Fixed, thanks!

**Review by wuxibin89 — APPROVED** [review]
